src/server/worker/datum_iterator.go

```diff
@@ -10,8 +10,13 @@ import (
 	"github.com/pachyderm/pachyderm/src/client"
 	"github.com/pachyderm/pachyderm/src/client/pfs"
 	"github.com/pachyderm/pachyderm/src/client/pps"
+
+	"github.com/cevaris/ordered_map"
 )
 
+// InputI is an interface for generic inputs
+type InputI interface{}
```

Comment:
The typing in this file smells a little weird to me and seems kind of over-engineered. I think most of it comes from needing two different ways to construct cross datum iterators, one is the way were before where you get a `[]*pps.Input` the other is where you get a `[][]*Input`. I think this code would simplify if you just had two constructors for the cross datum iterator, the old one and one that could be called directly by `newJoinDatumIterator`, that would look like:

```
func newCrossDatumIterator(cross [][]*Input)
```
Which could then call `newListDatumIterator` without having to do any reflection because it'd always have the write types to pass to it.